exporter/exporterhelper/wal.go

```diff
@@ -0,0 +1,208 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exporterhelper
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/jaegertracing/jaeger/pkg/queue"
+	"github.com/nsqio/go-diskqueue"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// WalQueue holds the WAL-backed queue
+type WalQueue struct {
+	logger      *zap.Logger
+	workers     int
+	stopWG      sync.WaitGroup
+	walMu       sync.Mutex
+	wal         diskqueue.Interface
+	stopped     *atomic.Uint32
+	quit        chan struct{}
+	unmarshaler requestUnmarshaler
+}
+
+func newWalQueue(logger *zap.Logger, path string, id string, syncFrequency time.Duration, unmarshaler requestUnmarshaler) *WalQueue {
+	wq, err := createWalQueue(logger, path, id, syncFrequency, unmarshaler)
+	if err != nil {
+		panic(err)
+	}
+	return wq
+}
+
+const (
+	// TODO: expose those as configuration parameters
+	maxBytesPerFile = int64(1 << 28)
+	minMsgSize      = int32(1)
+	maxMsgSize      = int32(1 << 27)
+	syncEvery       = int64(1000)
+)
+
+var (
+	errNotFound = errors.New("not found")
+)
+
+func dqLogOption(logger *zap.Logger, lvl diskqueue.LogLevel, f string, args ...interface{}) {
+	if logger == nil {
+		return
+	}
+
+	switch lvl {
+	case 1:
+		logger.Debug(fmt.Sprintf(f, args...))
+	case 2:
+		logger.Info(fmt.Sprintf(f, args...))
+	case 3:
+		logger.Warn(fmt.Sprintf(f, args...))
+	case 4:
+		logger.Error(fmt.Sprintf(f, args...))
+	case 5:
+		logger.Fatal(fmt.Sprintf(f, args...))
+	}
+}
+
+func createWalQueue(logger *zap.Logger, path string, id string, syncFrequency time.Duration, unmarshaler requestUnmarshaler) (*WalQueue, error) {
+	syncOption := syncFrequency == 0
+	syncEveryOption := syncEvery
+	if syncOption {
+		syncEveryOption = 1
+	}
+
+	walPath := filepath.Join(path, "wal")
+	err := os.Mkdir(walPath, 0700)
+	if err != nil && !os.IsExist(err) {
+		return nil, err
+	}
+
+	// TODO: sanitize id?
+	_wal := diskqueue.New(id, walPath,
+		maxBytesPerFile, minMsgSize, maxMsgSize, syncEveryOption, syncFrequency,
+		func(lvl diskqueue.LogLevel, f string, args ...interface{}) {
+			dqLogOption(logger, lvl, f, args)
+		})
+
+	wq := &WalQueue{
+		logger:      logger,
+		wal:         _wal,
+		unmarshaler: unmarshaler,
+		stopped:     atomic.NewUint32(0),
+		quit:        make(chan struct{}),
+	}
+
+	return wq, nil
+}
+
+// StartConsumers starts the given number of consumers which will be consuming items
```

Comment:
Can we document it a little bit more?
As far as I understand the implementation it blocks until `wq.stopped.Load() == 1`, can you describe that in a comment?
How should this communicate with `wq.quit` being close in `Stop()`? Is the intent that the above condition is reached by calling `wq.wal.Close()` in `close`?

Reply:
> As far as I understand the implementation it blocks until wq.stopped.Load() == 1, can you describe that in a comment?
No, it blocks only until all consumers (goroutines) are started

Reply:
Ah ok true that, I missed the `go func()` bit.

What about the remainder of my comment/question: Is that true? What does the communication between the "worker goroutine" and `close()` look like?

Reply:
Not sure if I am mising something, but regarding:

> What about the remainder of my comment/question: Is that true? What does the communication between the "worker goroutine" and close() look like?
> How should this communicate with wq.quit being close in Stop()? Is the intent that the above condition is reached by calling wq.wal.Close() in close?

Not exactly, it happens in `Stop()`:

```
func (wq *WALQueue) Stop() {
	wq.logger.Debug("Stopping WAL")
	wq.stopped.Store(1)
	wq.quitOnce.Do(func() {
		close(wq.quit)
	})
	wq.stopWG.Wait()
	err := wq.close()
	if err != nil {
		wq.logger.Error("Error when closing WAL", zap.Error(err))
	}
}

Reply:
Ok now I see, you're using `wq.stopped` atomic. Alright.

So: just one more comment: you should be able to use `wg.quit`since it's closed in `Stop()` anyway


Something like:

```go
select {
    case <-wg.quit:
        return
    default:
}
```